daemon/api_validate.go

```diff
@@ -264,38 +292,64 @@ func applyValidationSet(c *Command, r *http.Request, _ *auth.UserState) Response
 	case "forget":
 		return forgetValidationSet(st, accountID, name, req.Sequence)
 	case "apply":
-		return updateValidationSet(st, accountID, name, req.Mode, req.Sequence)
+		return updateValidationSet(st, accountID, name, req.Mode, req.Sequence, user)
 	default:
 		return BadRequest("unsupported action %q", req.Action)
 	}
 }
 
-func updateValidationSet(st *state.State, accountID, name string, reqMode string, sequence int) Response {
+// updateValidationSet handles snap validate --monitor and --enforce accountId/name[=sequence].
+func updateValidationSet(st *state.State, accountID, name string, reqMode string, sequence int, user *auth.UserState) Response {
 	var mode assertstate.ValidationSetMode
+	// TODO: only monitor mode for now, add enforce.
 	switch reqMode {
 	case "monitor":
 		mode = assertstate.Monitor
-	case "enforce":
-		mode = assertstate.Enforce
 	default:
 		return BadRequest("invalid mode %q", reqMode)
 	}
 
-	// TODO: if pinned, check if we have the needed assertion locally;
-	// check with the store if there is something newer there;
-	// check what is the latest in the store if the assertion is not pinned.
+	var err error
+	var as *asserts.ValidationSet
+
+	pinned := sequence > 0
+	if pinned {
+		st.Unlock()
+		as, err = validationSetAssertFromDb(st, accountID, name, sequence)
+		st.Lock()
+		if err != nil {
+			if _, ok := err.(*asserts.NotFoundError); !ok {
+				return InternalError(err.Error())
+			}
+		}
+	}
+
+	// not pinned or not found locally, get from the store
+	if !pinned || as == nil {
+		// sequence == 0 (i.e. not pinned) gets the latest from the store
+		storeAs, err := getSingleSeqFormingAssertion(st, accountID, name, sequence, user)
```

Comment:
@pedronis this is the bit I was mentioning in the standup. Maybe we should use pool to fetch also developer key etc.

Reply:
yes, now I see what you asked, let's chat about it. Indeed this will not work in general case as is, because for monitor we should also store the assertion in the assertion db as well and that needs the key that signed it which we might not have